llama-index-core/llama_index/core/indices/property_graph/transformations/schema_llm.py

```diff
@@ -230,11 +206,17 @@ def validate(v: Any, values: Any) -> Any:
             )
             kg_schema_cls.__doc__ = "Knowledge Graph Schema."
 
+        # Get validation schema
+        kg_validation_schema = kg_validation_schema or DEFAULT_VALIDATION_SCHEMA
+        # TODO: Remove this in a future version & encourage List[Triple] for validation schema
+        if isinstance(kg_validation_schema, list):
+            kg_validation_schema = {"relationships": kg_validation_schema}
```

Comment:
Hmm, we don't need to update the `def _prune_invalide_triplets()` function at all? Seems like it would need to be updated so that it works for a list of triples?

Reply:
Hmm, interesting that it even worked with this change as is. I'll go into that method deeper and see how to make it actually do what it's supposed to, lol.

Reply:
I wonder if the schema validation dict/list is just being taken into the prompt as a text field without any actual validation?

Reply:
OK, so I added some additional checks to verify if a provided relation isn't to be accepted as per the list of triples.

## Test 1:

```py
entities = Literal["PERSON", "ORGANIZATION", "LOCATION"]
relations = Literal["IS_CEO_OF", "IN_LOCATION"]

# Missing the "IN_LOCATION" relationship
validation_schema = [
    ("PERSON", "IS_CEO_OF", "ORGANIZATION"),
]
```

It works as intended, and only returns this graph:

![image](https://github.com/run-llama/llama_index/assets/35005448/c3012726-5e7e-4bcd-8465-362f1eeb1a47)

## Test 2:

```py
entities = Literal["PERSON", "ORGANIZATION", "LOCATION"]
relations = Literal["IS_CEO_OF", "IN_LOCATION"]

# Non-existent relationship type "ABC" is specified instead of "IN_LOCATION"
validation_schema = [
    ("PERSON", "IS_CEO_OF", "ORGANIZATION"),
    ("ORGANIZATION", "ABC", "LOCATION"),
]
```

When a non-existent relationship is present, it correctly skips the relationship and returns the correct graph as above.

The if-else logic isn't ideal but it at least allows the existing blog posts and other stuff out there to remain valid. Hope this makes sense!

Reply:
_prune_invalid_triplets needs updated to check for valid triplets (original method doesn't prune correctly anymore):

```shell
    def _prune_invalid_triplets(self, kg_schema: Any) -> List[Tuple[EntityNode, Relation, EntityNode]]:
        """Prune invalid triplets."""
        assert isinstance(kg_schema, self.kg_schema_cls)

        valid_triplets = []
        for triplet in kg_schema.triplets:
            subject = triplet.subject.name
            subject_type = triplet.subject.type

            relation = triplet.relation.type

            obj = triplet.object.name
            obj_type = triplet.object.type

            print (self.kg_validation_schema['relationships'])
            # Check if the triplet is valid
            if (subject_type, relation, obj_type) not in self.kg_validation_schema['relationships']:
                continue

            # Remove self-references
            if subject.lower() == obj.lower():
                continue

            subj_node = EntityNode(label=subject_type, name=subject)
            obj_node = EntityNode(label=obj_type, name=obj)
            rel_node = Relation(
                label=relation, source_id=subj_node.id, target_id=obj_node.id
            )
            valid_triplets.append((subj_node, rel_node, obj_node))

        return valid_triplets
```

Reply:
Also the validation schema should really be in dictionary form here for O(1) lookup. Something like this:

DEFAULT_VALIDATION_SCHEMA = {
    "PRODUCT": {
        "USED_BY": {"PRODUCT"},
        "USED_FOR": {"MARKET"},
        "HAS": {"TECHNOLOGY"}
    },
    "MARKET": {
        "LOCATED_IN": {"LOCATION"},
        "HAS": {"TECHNOLOGY"}
    },
    "TECHNOLOGY": {
        "USED_BY": {"PRODUCT"},
        "USED_FOR": {"MARKET"},
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"ORGANIZATION"},
        "IS_A": {"PRODUCT"}
    },
    "EVENT": {
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"ORGANIZATION"}
    },
    "CONCEPT": {
        "USED_BY": {"TECHNOLOGY"},
        "USED_FOR": {"PRODUCT"}
    },
    "ORGANIZATION": {
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"ORGANIZATION", "MARKET"}
    },
    "PERSON": {
        "BORN_IN": {"LOCATION", "TIME"},
        "DIED_IN": {"LOCATION", "TIME"},
        "WORKED_ON": {"EVENT", "PRODUCT", "CONCEPT", "TECHNOLOGY"}
    },
    "LOCATION": {
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"LOCATION"}
    }
}

Reply:
@mphipps2 did you try the code as per https://github.com/run-llama/llama_index/pull/14357/commits/8fbf8bff49ef80efc26a484935087fbf127fb8c5?

Reply:
I'd actually recommend copy-pasting the entire contents of `schema_llm.py` that I have in this PR to the LlamaIndex code directory in your virtual environment to test if that helps your scenario.

Reply:
Yep, tested that and it works for me too. I'll make the update.

Reply:
Per the latest commit in 65be6ce7dabfcd8d600db4ae77c093fb8bcf4cce, here's an example of the print status if a relationship is skipped or invalid:

```py
# Intentionally skipping mentioning the `IS_CEO_OF` triplet
validation_schema = [
    ("ORGANIZATION", "IN_LOCATION", "LOCATION"),
]
```

Output:

```
Parsing nodes: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 1426.15it/s]
Extracting paths from text with schema:   0%|                                                                                                                                                            | 0/1 [00:00<?, ?it/s]
The following triplet is skipped as it isn't in the validation schema: (PERSON, IS_CEO_OF, ORGANIZATION)
Extracting paths from text with schema: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.61s/it]
Generating embeddings: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  5.03it/s]
Generating embeddings: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  4.97it/s]
```

The output graph turns out without this relationship.

Reply:
This nested validation structure in the version below will be easier to use (tested for current format and backward compatibility):

```shell

import asyncio
from typing import Any, Dict, List, Literal, Optional, Tuple, Union, Set

try:
    from typing import TypeAlias
except ImportError:
    # python 3.8 and 3.9 compatibility
    TypeAlias = Any

from llama_index.core.async_utils import run_jobs
from llama_index.core.bridge.pydantic import create_model, validator, Field
from llama_index.core.graph_stores.types import (
    EntityNode,
    Relation,
    Triplet,
    KG_NODES_KEY,
    KG_RELATIONS_KEY,
)
from llama_index.core.prompts import PromptTemplate
from llama_index.core.schema import TransformComponent, BaseNode
from llama_index.core.llms.llm import LLM


DEFAULT_ENTITIES = Literal[
    "PRODUCT",
    "MARKET",
    "TECHNOLOGY",
    "EVENT",
    "CONCEPT",
    "ORGANIZATION",
    "PERSON",
    "LOCATION",
    "TIME",
    "MISCELLANEOUS",
]

DEFAULT_RELATIONS = Literal[
    "USED_BY",
    "USED_FOR",
    "LOCATED_IN",
    "PART_OF",
    "WORKED_ON",
    "HAS",
    "IS_A",
    "BORN_IN",
    "DIED_IN",
    "HAS_ALIAS",
]

DEFAULT_VALIDATION_SCHEMA: Dict[str, Dict[str, Set[str]]] = {
    "PRODUCT": {
        "USED_BY": {"PRODUCT"},
        "USED_FOR": {"MARKET"},
        "HAS": {"TECHNOLOGY"}
    },
    "MARKET": {
        "LOCATED_IN": {"LOCATION"},
        "HAS": {"TECHNOLOGY"}
    },
    "TECHNOLOGY": {
        "USED_BY": {"PRODUCT"},
        "USED_FOR": {"MARKET"},
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"ORGANIZATION"},
        "IS_A": {"PRODUCT"}
    },
    "EVENT": {
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"ORGANIZATION"}
    },
    "CONCEPT": {
        "USED_BY": {"TECHNOLOGY"},
        "USED_FOR": {"PRODUCT"}
    },
    "ORGANIZATION": {
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"ORGANIZATION", "MARKET"}
    },
    "PERSON": {
        "BORN_IN": {"LOCATION", "TIME"},
        "DIED_IN": {"LOCATION", "TIME"},
        "WORKED_ON": {"EVENT", "PRODUCT", "CONCEPT", "TECHNOLOGY"}
    },
    "LOCATION": {
        "LOCATED_IN": {"LOCATION"},
        "PART_OF": {"LOCATION"}
    }
}

DEFAULT_SCHEMA_PATH_EXTRACT_PROMPT = PromptTemplate(
    "Give the following text, extract the knowledge graph according to the provided schema. "
    "Try to limit to the output {max_triplets_per_chunk} extracted paths.s\n"
    "-------\n"
    "{text}\n"
    "-------\n"
)


class SchemaLLMPathExtractor(TransformComponent):
    """
    Extract paths from a graph using a schema.

    Args:
        llm (LLM):
            The language model to use.
        extract_prompt (Union[PromptTemplate, str], optional):
            The template to use for the extraction query. Defaults to None.
        possible_entities (Optional[TypeAlias], optional):
            The possible entities to extract. Defaults to None.
        possible_relations (Optional[TypeAlias], optional):
            The possible relations to extract. Defaults to None.
        strict (bool, optional):
            Whether to enforce strict validation of entities and relations. Defaults to True.
            If false, values outside of the schema will be allowed.
        kg_schema_cls (Any, optional):
            The schema class to use. Defaults to None.
        kg_validation_schema (Dict[str, Any], optional):
            The validation schema to use. Defaults to None.
        max_triplets_per_chunk (int, optional):
            The maximum number of triplets to extract per chunk. Defaults to 10.
        num_workers (int, optional):
            The number of workers to use. Defaults to 4.
    """

    llm: LLM
    extract_prompt: PromptTemplate
    kg_schema_cls: Any
    kg_validation_schema: Dict[str, Any]
    num_workers: int
    max_triplets_per_chunk: int
    strict: bool

    def __init__(
        self,
        llm: LLM,
        extract_prompt: Union[PromptTemplate, str] = None,
        possible_entities: Optional[TypeAlias] = None,
        possible_relations: Optional[TypeAlias] = None,
        strict: bool = True,
        kg_schema_cls: Any = None,
        kg_validation_schema: Dict[str, Any] = None,
        max_triplets_per_chunk: int = 10,
        num_workers: int = 4,
    ) -> None:
        """Init params."""
        if isinstance(extract_prompt, str):
            extract_prompt = PromptTemplate(extract_prompt)

        # Build a pydantic model on the fly
        if kg_schema_cls is None:
            possible_entities = possible_entities or DEFAULT_ENTITIES
            entity_cls = create_model(
                "Entity",
                type=(
                    possible_entities if strict else str,
                    Field(
                        ...,
                        description=(
                            "Entity in a knowledge graph. Only extract entities with types that are listed as valid: "
                            + str(possible_entities)
                        ),
                    ),
                ),
                name=(str, ...),
            )

            possible_relations = possible_relations or DEFAULT_RELATIONS
            relation_cls = create_model(
                "Relation",
                type=(
                    possible_relations if strict else str,
                    Field(
                        ...,
                        description=(
                            "Relation in a knowledge graph. Only extract relations with types that are listed as valid: "
                            + str(possible_relations)
                        ),
                    ),
                ),
            )

            triplet_cls = create_model(
                "Triplet",
                subject=(entity_cls, ...),
                relation=(relation_cls, ...),
                object=(entity_cls, ...),
            )

            def validate(v: Any, values: Any) -> Any:
                """Validate triplets."""
                passing_triplets = []
                for i, triplet in enumerate(v):
                    # cleanup
                    try:
                        for key in triplet:
                            triplet[key]["type"] = triplet[key]["type"].replace(
                                " ", "_"
                            )
                            triplet[key]["type"] = triplet[key]["type"].upper()

                        # validate, skip if invalid
                        _ = triplet_cls(**triplet)
                        passing_triplets.append(v[i])
                    except (KeyError, ValueError):
                        continue

                return passing_triplets

            root = validator("triplets", pre=True)(validate)
            kg_schema_cls = create_model(
                "KGSchema",
                __validators__={"validator1": root},
                triplets=(List[triplet_cls], ...),
            )
            kg_schema_cls.__doc__ = "Knowledge Graph Schema."

        # Get validation schema
        kg_validation_schema = kg_validation_schema or DEFAULT_VALIDATION_SCHEMA

        super().__init__(
            llm=llm,
            extract_prompt=extract_prompt or DEFAULT_SCHEMA_PATH_EXTRACT_PROMPT,
            kg_schema_cls=kg_schema_cls,
            kg_validation_schema=kg_validation_schema,
            num_workers=num_workers,
            max_triplets_per_chunk=max_triplets_per_chunk,
            strict=strict,
        )

    @classmethod
    def class_name(cls) -> str:
        return "SchemaLLMPathExtractor"

    def __call__(
        self, nodes: List[BaseNode], show_progress: bool = False, **kwargs: Any
    ) -> List[BaseNode]:
        """Extract triplets from nodes."""
        return asyncio.run(self.acall(nodes, show_progress=show_progress, **kwargs))

    def _prune_invalid_triplets(self, kg_schema: Any) -> List[Tuple[EntityNode, Relation, EntityNode]]:
        """Prune invalid triplets."""
        assert isinstance(kg_schema, self.kg_schema_cls)

        valid_triplets = []
        for triplet in kg_schema.triplets:
            subject = triplet.subject.name
            subject_type = triplet.subject.type

            relation = triplet.relation.type

            obj = triplet.object.name
            obj_type = triplet.object.type

            # Check if the triplet is valid based on the schema format
            if isinstance(self.kg_validation_schema, dict):
                # Check if the schema is in the new dictionary format
                if any(isinstance(v, dict) for v in self.kg_validation_schema.values()):
                    # New schema format
                    if subject_type not in self.kg_validation_schema or \
                    relation not in self.kg_validation_schema[subject_type] or \
                    obj_type not in self.kg_validation_schema[subject_type][relation]:
                        continue
                else:
                    # Original schema format
                    if relation not in self.kg_validation_schema.get(subject_type, []) and \
                    relation not in self.kg_validation_schema.get(obj_type, []):
                        continue
            else:
                # Handle invalid schema format case
                print("Invalid schema format")
                continue

            # Remove self-references
            if subject.lower() == obj.lower():
                continue

            subj_node = EntityNode(label=subject_type, name=subject)
            obj_node = EntityNode(label=obj_type, name=obj)
            rel_node = Relation(
                label=relation, source_id=subj_node.id, target_id=obj_node.id
            )
            valid_triplets.append((subj_node, rel_node, obj_node))

        return valid_triplets


    async def _aextract(self, node: BaseNode) -> BaseNode:
        """Extract triplets from a node."""
        assert hasattr(node, "text")

        text = node.get_content(metadata_mode="llm")
        try:
            kg_schema = await self.llm.astructured_predict(
                self.kg_schema_cls,
                self.extract_prompt,
                text=text,
                max_triplets_per_chunk=self.max_triplets_per_chunk,
            )
            triplets = self._prune_invalid_triplets(kg_schema)
        except ValueError:
            triplets = []

        existing_nodes = node.metadata.pop(KG_NODES_KEY, [])
        existing_relations = node.metadata.pop(KG_RELATIONS_KEY, [])

        metadata = node.metadata.copy()
        for subj, rel, obj in triplets:
            subj.properties = metadata
            obj.properties = metadata
            rel.properties = metadata

            existing_relations.append(rel)
            existing_nodes.append(subj)
            existing_nodes.append(obj)

        node.metadata[KG_NODES_KEY] = existing_nodes
        node.metadata[KG_RELATIONS_KEY] = existing_relations

        return node

    async def acall(
        self, nodes: List[BaseNode], show_progress: bool = False, **kwargs: Any
    ) -> List[BaseNode]:
        """Extract triplets from nodes async."""
        jobs = []
        for node in nodes:
            jobs.append(self._aextract(node))

        return await run_jobs(
            jobs,
            workers=self.num_workers,
            show_progress=show_progress,
            desc="Extracting paths from text with schema",
        )
```

Reply:
@mphipps2 I don't see why O(1) lookups would be an issue here that necessitates a dictionary because the list of relationship triples is finite (typically under 100, for nearly all use cases), so that's a very small lookup. Also, I'd argue (and @logan-markewich might agree) that this isn't really very readable and easy for a user to enter. Having ~100 triples explicitly defined in the script or in an external text file is very clean, explicit, and can be sorted alphabetically for a user to easily find the triple they're looking for.

Also, I'm not sure why the value type of each relationship is a singleton set, for e.g., `{"PRODUCT"}` in the below example. It's really far less readable and harder to manage from a user perspective imo.

```
    "PRODUCT": {
        "USED_BY": {"PRODUCT"},
        "USED_FOR": {"MARKET"},
        "HAS": {"TECHNOLOGY"}
    },
```

Reply:
Maybe just personal preference. Difference is really only when the number of valid triplets is large.

Either way though it should work.

Reply:
I think I agree that a flatter list is better dev UX. Even if it was 1000 triples, the lookup would still be fast.